[WIP] Provide a working setup

fosskers opened the pull request:
### TODO
- [x] `docker-compose` config
- [x] Vagrant config
- [x] Terraform config
- [x] Scripts
- [x] README
- [x] Testing
- [ ] docker images to quay
- [ ] Deal with feedback

### Motivation
This brings our website and demo deployment under one unified configuration. All demos and their API calls are also managed by a single nginx, with Transit now available under `http://<host>/transit/` and Chatta under `http://<host>/chatta/`.

Further long-lived demos should be added in a similar way.

fosskers committed:
Readd files

fosskers committed:
Readd submodules

hectcastro commented on .gitmodules:
```diff
@@ -0,0 +1,9 @@
+[submodule "geotrellis-site"]
```
Would it make sense to just reference container image IDs from this project? Something like:

```bash
$ GT_SITE_REV="..." GT_TRANSIT_REV="..." GT_CHATTA_REV="..." ./scripts/infra plan
```

This would allow us to avoid establishing a direct connection between this repository and the others. It would also make deployments through a CI tool more flexible.

fosskers replied:
Potentially, although that would break current `docker-compose build` behaviour.

fosskers replied:
I do think that submodules overall are a bit of a pain.

hectcastro commented on Makefile:
```diff
@@ -0,0 +1,49 @@
+	cd geotrellis-transit && ./sbt clean
+
+# One-time data fetches/ingests
+setup: chatta
```
I think that it would valuable to move container image creation and publishing out to the project repositories vs. doing it all here. Presumably, those container images would have value on their own, and it would be good to release them at an interval that makes sense for them in isolation.

fosskers replied:
So long as keeping track of the individual commit hashes isn't more pain than it's worth. I'm still pondering this.

hectcastro replied:
Thus far, this particular situation is a little uncommon. The closest example we have is RF.

With RF, there is a repository that contains all of the services required to drive the application, but all of the Terraform configuration is in another repository:

- https://github.com/azavea/raster-foundry
- https://github.com/azavea/raster-foundry-deployment

The high level approach is that all of the artifacts required to drive a deployment are created as part of the CI process for the main repository (container images, JARs, etc.):

- https://github.com/azavea/raster-foundry/blob/develop/Jenkinsfile#L10-L36

From there, the deployment repository is pulled down after tests past for certain branches, and a new deployment is kicked off:

- https://github.com/azavea/raster-foundry/blob/develop/Jenkinsfile#L38-L64

`GIT_COMMIT` is passed to `infra`, which takes care of threading the value of that variable as the container image tag provided to ECS:

- https://github.com/azavea/raster-foundry-deployment/blob/develop/scripts/infra#L68

The flexibility of using `GIT_COMMIT` comes into play when we need to deploy old versions of the application (to reproduce a bug), new versions (test a new feature in AWS), or just different environments (staging vs. production). The main requirement is that images already exist in the container image registries with references to `GIT_COMMIT`.

hectcastro commented on docker-compose.yml:
```diff
@@ -0,0 +1,44 @@
+services:
+
+  gtsite-service:
+    image: fosskers/gtsite-service
```
This should be renamed to something generic for the project. Possibly just `gtsite-service` or similar.

We can re-tag container images locally so that their names correspond to the remote registry:

```bash
docker tag "raster-foundry-nginx:${GIT_COMMIT}" \
       "${AWS_ECR_ENDPOINT}/raster-foundry-nginx:${GIT_COMMIT}"
docker tag "raster-foundry-app-server:${GIT_COMMIT}" \
       "${AWS_ECR_ENDPOINT}/raster-foundry-app-server:${GIT_COMMIT}"
docker tag "raster-foundry-tile-server:${GIT_COMMIT}" \
       "${AWS_ECR_ENDPOINT}/raster-foundry-tile-server:${GIT_COMMIT}"
docker tag "raster-foundry-airflow:${GIT_COMMIT}" \
       "${AWS_ECR_ENDPOINT}/raster-foundry-airflow:${GIT_COMMIT}"
docker tag "raster-foundry-app-migrations:${GIT_COMMIT}" \
       "${AWS_ECR_ENDPOINT}/raster-foundry-migrations:${GIT_COMMIT}"
```

fosskers replied:
The previous config (just for the site, no demos) did something similar to what I'm doing here, except it tagged the images as `quay.io/geotrellis/foobar`.

hectcastro commented on scripts/cipublish:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
I think that we need to include something here to put in place whatever credentials are necessary for pushing to the remote container image registry. For the EC2 Container Registry, it looks like this:

```bash
# Evaluate the return value of the get-login subcommand, which
# is a docker login command with temporarily ECR credentials.
eval "$(aws ecr get-login)"
```

This also assumes that `AWS_PROFILE` is setup so that the AWS CLI can authenticate with the AWS API.

fosskers replied:
The publishing story needs to be improved in general: at the moment the images are publishing to my Docker Hub.

hectcastro replied:
The existing story around publishing thin OpenJDK 8 container images wrapping an OpenJDK 8 built JAR is defined here (the ECR part can be altered to support Quay, Docker Hub, etc.):

- https://github.com/azavea/raster-foundry/blob/develop/scripts/cipublish#L36-L73
- https://github.com/azavea/raster-foundry/blob/develop/app-backend/app/Dockerfile
- https://github.com/azavea/raster-foundry/blob/develop/app-backend/tile/.dockerignore

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+    "image": "fosskers/gtsite-service",
+    "memory": 1024,
+    "essential": true,
+    "portMappings": [
```
There port mappings aren't necessary. Interaction with the `gtsite-service` will happen via the link established by the Nginx service.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+  {
+    "name": "gtsite-service",
+    "image": "fosskers/gtsite-service",
+    "memory": 1024,
```
We should include an override for `command` somewhere in this definition to capture setting `-Xmx`:

```json
command": [
    "-Xmx768m",
    "-jar",
    "something.jar"
],
```

**Note**: This explicitly sets the maximum JVM heap so that it fits within the overall 1GB container memory allocation. Also, this assumes that `java` is the `entrypoint` for the underlying container image.

fosskers replied:
I'll consult the config from that NDA'd project we previously collaborated on.

hectcastro replied:
I think the most up-to-date example might be:

- https://github.com/azavea/raster-foundry/blob/develop/app-backend/app/Dockerfile
- https://github.com/azavea/raster-foundry-deployment/blob/develop/terraform/task-definitions/app-server.json#L43-L47

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+  {
+    "name": "gt-transit",
+    "image": "fosskers/gt-transit",
+    "memory": 3072,
```
Same comments as above with regard to the explicit upper bound of the JVM heap via `command`. In this case, we probably want a ~2.5GB heap.

fosskers replied:
2.5 may crash the JVM, I've seen `transit` get pretty thirsty during operation.

hectcastro replied:
If the JVM heap needs 3GB, then we just need to make sure that `memory` is at around 3.5GB, otherwise you'll run the risk of OOM.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+        "containerPort": 8080,
+        "hostPort": 8080
+      }
+    ]
```
Similar comment here regarding a `logConfiguration`.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+    "image": "fosskers/gt-transit",
+    "memory": 3072,
+    "essential": true,
+    "portMappings": [
```
Similar comment as about about omitting this `portMappings` key.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+        "containerPort": 9999,
+        "hostPort": 9999
+      }
+    ]
```
Similar comment as above about adding a `logConfiguration`.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+  {
+    "name": "gt-chatta-demo",
+    "image": "fosskers/gt-chatta-demo",
+    "memory": 1024,
```
Similar comment as above with regard to setting `-Xmx` via `command`. In this case, a 768MB heap.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+    "memory": 1024,
+    "essential": false,
+    "hostname": "chatta",
+    "portMappings": [
```
Similar comment as above about removing `portMappings`.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+        "containerPort": 8777,
+        "hostPort": 8777
+      }
+    ]
```
Similar comment as above about adding a `logConfiguration`.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+provider "aws" {
+  profile = "${var.profile}" # Assumes you have valid AWS credentials.
```
We should not have to include this setting here. Instead, we should be able to supply `AWS_PROFILE` when invoking the `terraform` command.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+resource "aws_ecs_cluster" "gt-site-cluster" {
+  name = "gt-site-cluster"
+
+  lifecycle {
```
We should remove this `lifecycle` setting from everywhere except the `aws_launch_configuration` resource.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+# --- ECS RESOURCES --- #
+
+resource "aws_ecs_cluster" "gt-site-cluster" {
+  name = "gt-site-cluster"
```
For naming AWS resource, I think that it would be good to follow the existing camel case convention. See the link below for examples. In this case, something like `ecs${var.environment}Cluster`.

- See: https://github.com/azavea/raster-foundry-deployment/

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+# `containers.json` is analogous to `docker-compose.yml`.
+# All the container settings are there.
+resource "aws_ecs_task_definition" "gt-site-cluster-task" {
+  family                = "gt-site-cluster"
```
RE: Naming; something like `${var.environment}GTSite`.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+# All the container settings are there.
+resource "aws_ecs_task_definition" "gt-site-cluster-task" {
+  family                = "gt-site-cluster"
+  container_definitions = "${file("containers.json")}"
```
We should use `containers.json` as a template that can be optionally rendered with Terraform variables:

```hcl
data "template_file" "app_server_http_ecs_task" {
  template = "${file("task-definitions/nginx.json")}"

  vars {
     ...
  }
}
```

Then `container_definitions` ends up looking like:

```hcl
container_definitions = "${data.template_file.app_server_http_ecs_task.rendered}"
```

fosskers replied:
I had seen this elsewhere but thought it unnecessary for this case.

hectcastro replied:
As soon as you add the `logConfiguration`, you'll have to thread in the AWS region for `awslogs-region`. At that point, you'll have to decide between using a template or hard coding `us-east-1`.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+}
+
+resource "aws_ecs_service" "gt-site-cluster-service" {
+  name                               = "gt-site-cluster"
```
RE: Naming; something like `${var.environment}GTSite`.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  cluster                            = "${aws_ecs_cluster.gt-site-cluster.id}"
+  task_definition                    = "${aws_ecs_task_definition.gt-site-cluster-task.family}:${aws_ecs_task_definition.gt-site-cluster-task.revision}"
+  desired_count                      = "${var.desired_instance_count}"
+  iam_role                           = "${var.ecs_service_role}"
```
We should be able to create the IAM resources necessary via Terraform as well. Something like:

```hcl
data "aws_iam_policy_document" "container_instance_ecs_assume_role" {
  statement {
    effect = "Allow"

    principals {
      type        = "Service"
      identifiers = ["ecs.amazonaws.com"]
    }

    actions = ["sts:AssumeRole"]
  }
}

resource "aws_iam_role" "container_instance_ecs" {
  name               = "ecs${var.environment}InstanceRole"
  assume_role_policy = "${data.aws_iam_policy_document.container_instance_ecs_assume_role.json}"
}

resource "aws_iam_role_policy_attachment" "ecs_for_ec2_policy_app_server_ecs_role" {
  role       = "${aws_iam_role.container_instance_ecs.name}"
  policy_arn = "${var.aws_ecs_for_ec2_service_role_policy_arn}"
}

resource "aws_iam_role_policy_attachment" "ecs_policy" {
  role       = "${aws_iam_role.container_instance_ecs.name}"
  policy_arn = "${var.aws_ecs_service_role_policy_arn}"
}
```

Here, `aws_ecs_for_ec2_service_role_policy_arn` and `aws_ecs_service_role_policy_arn` are ARNs to canned policies:

- `arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role`
- `arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceRole`

fosskers replied:
These were copy-pasted from other projects since they were known to work. I've had bad luck with trying to create new IAM resources in the past.

hectcastro replied:
A concern is that changes made to the role to satisfy the copy-pasted project's purposes will automatically cascade to this one (either escalating privileges, or taking them away and causing breakage). In addition, `ecs_instance_profile` resolves to another Terraform managed resource.

Codifying the IAM resources used by each project in Terraform is preferred.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  task_definition                    = "${aws_ecs_task_definition.gt-site-cluster-task.family}:${aws_ecs_task_definition.gt-site-cluster-task.revision}"
+  desired_count                      = "${var.desired_instance_count}"
+  iam_role                           = "${var.ecs_service_role}"
+  deployment_minimum_healthy_percent = "0"
```
We should make these two settings parameters. In order to avoid downtime, we'll also have to set minimum to 100 and maximum to 200.

fosskers replied:
I didn't know what would be good values for these, thanks.

hectcastro commented on terraform/containers.json:
```diff
@@ -0,0 +1,56 @@
+    "portMappings": [
+      {
+        "containerPort": 8080,
+        "hostPort": 8080
```
`hostPort` here should be set to 0 so that ECS can handle dynamic port mappings at the target group.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+}
+
+# A load balancer is necessary to launch the EC2 manager instance properly.
+resource "aws_elb" "gt-site-elb" {
```
We should be using an `aws_alb` resources vs. the legacy `aws_elb`. The set of resources necessary to make that happen look something like:

```hcl
resource "aws_alb" "app_server" {
  security_groups = ["${aws_security_group.app_server_alb.id}"]
  subnets         = ["${module.vpc.public_subnet_ids}"]
  name            = "alb${var.environment}AppServer"

  tags {
    Name        = "albAppServer"
    Project     = "${var.project}"
    Environment = "${var.environment}"
  }
}

resource "aws_alb_target_group" "app_server_https" {
  # Name can only be 32 characters long, so we MD5 hash the name and
  # truncate it to fit.
  name = "tf-tg-${replace("${md5("${var.environment}HTTPSAppServer")}", "/(.{0,26})(.*)/", "$1")}"

  health_check {
    healthy_threshold   = "3"
    interval            = "30"
    protocol            = "HTTP"
    timeout             = "3"
    path                = "/healthcheck/"
    unhealthy_threshold = "2"
  }

  port     = "443"
  protocol = "HTTP"
  vpc_id   = "${module.vpc.id}"

  tags {
    Name        = "tg${var.environment}HTTPSAppServer"
    Project     = "${var.project}"
    Environment = "${var.environment}"
  }
}

resource "aws_alb_listener" "app_server_https" {
  load_balancer_arn = "${aws_alb.app_server.id}"
  port              = "443"
  protocol          = "HTTPS"
  certificate_arn   = "${var.ssl_certificate_arn}"

  default_action {
    target_group_arn = "${aws_alb_target_group.app_server_https.id}"
    type             = "forward"
  }
}
```

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  deployment_maximum_percent         = "100"
+
+  load_balancer {
+    elb_name       = "${aws_elb.gt-site-elb.name}"
```
When the ALB is used, this setting gets replaced with:

```hcl
target_group_arn = "${aws_alb_target_group.app_server_https.id}"
```

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  }
+}
+
+resource "aws_launch_configuration" "ecs" {
```
This resource should have a `create_before_destroy` `lifecycle` setting.

hectcastro replied:
This resource also appears to be missing a `security_groups` attribute:

```hcl
security_groups      = ["${aws_security_group.container_instance.id}"]
```

The security group resource should look like:

```hcl
resource "aws_security_group" "container_instance" {
  vpc_id = "${module.vpc.id}"

  tags {
    Name        = "sgContainerInstance"
    Project     = "${var.project}"
    Environment = "${var.environment}"
  }
}
```

At the very least, you'll also want to add the following rule so that the ALB can talk to instance over any port (to support dynamic port mapping):

```hcl
resource "aws_security_group_rule" "alb_app_server_container_instance_all_egress" {
  type      = "egress"
  from_port = 0
  to_port   = 65535
  protocol  = "tcp"

  security_group_id        = "${aws_security_group.app_server_alb.id}"
  source_security_group_id = "${aws_security_group.container_instance.id}"
}
```

fosskers replied:
From the Terraform docs:
```
vpc_id - (Optional) The id of the VPC that the desired security group belongs to.
```
Where do we find this value, exactly? It's probably also dependent on whether we decide to create a new VPC for this.

fosskers replied:
```hlc
  security_group_id        = "${aws_security_group.app_server_alb.id}"
  source_security_group_id = "${aws_security_group.container_instance.id}"
```
Two security groups are referenced here, but the example above only contains one. Where does the other come from?

hectcastro replied:
Reusing the default VPC, the value for `vpc_id` would be `vpc-617f9604`.

Despite being marked as optional, that isn't realistically true. Prior to the existence of VPCs there was what is now known as "EC2-Classic". This environment gave each EC2 instance a public IP (and you had no other choice).

Not supplying `vpc_id` for a security group makes it so that the security group is an EC2-Classic security group, which isn't compatible with EC2-VPC security groups.

---

The other security group ID is referencing the security group for the load balancer itself. That's something like: https://github.com/azavea/raster-foundry-deployment/blob/develop/terraform/app.tf#L4-L12

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  name                 = "ECS ${aws_ecs_cluster.gt-site-cluster.name}"
+  image_id             = "${var.aws_ecs_ami}"
+  instance_type        = "${var.ec2_instance_type}"
+  iam_instance_profile = "${var.ecs_instance_profile}"
```
Similar to the ECS Service IAM role, this instance profile IAM resource should be created within Terraform:

```hcl
data "aws_iam_policy_document" "container_instance_ec2_assume_role" {
  statement {
    effect = "Allow"

    principals {
      type        = "Service"
      identifiers = ["ec2.amazonaws.com"]
    }

    actions = ["sts:AssumeRole"]
  }
}

resource "aws_iam_role" "container_instance_ec2" {
  name               = "${var.environment}ContainerInstanceProfile"
  assume_role_policy = "${data.aws_iam_policy_document.container_instance_ec2_assume_role.json}"
}

resource "aws_iam_role_policy_attachment" "ecs_for_ec2_policy_container_instance_role" {
  role       = "${aws_iam_role.container_instance_ec2.name}"
  policy_arn = "${var.aws_ecs_for_ec2_service_role_policy_arn}"
}

resource "aws_iam_role_policy_attachment" "cloudwatch_logs_policy_container_instance_role" {
  role       = "${aws_iam_role.container_instance_ec2.name}"
  policy_arn = "${var.aws_cloudwatch_logs_policy_arn}"
}

resource "aws_iam_instance_profile" "container_instance" {
  name  = "${aws_iam_role.container_instance_ec2.name}"
  roles = ["${aws_iam_role.container_instance_ec2.name}"]
}
```

Again, here `aws_ecs_for_ec2_service_role_policy_arn` and `aws_cloudwatch_logs_policy_arn` end up being the following values:

- `arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role`
- `arn:aws:iam::aws:policy/CloudWatchLogsFullAccess`

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  iam_instance_profile = "${var.ecs_instance_profile}"
+
+  key_name                    = "${var.ec2_key}"
+  associate_public_ip_address = true
```
We should not launch these container instances in public subnets, and they should not get public IPs. The only public endpoint to the services should be through the ALB.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+
+  key_name                    = "${var.ec2_key}"
+  associate_public_ip_address = true
+  user_data                   = "#!/bin/bash\necho ECS_CLUSTER='${aws_ecs_cluster.gt-site-cluster.name}' > /etc/ecs/ecs.config"
```
Similar to the ECS task definition, this value should be a rendered template:

```hcl
data "template_file" "container_instance_cloud_config" {
  template = "${file("cloud-config/container-instance.yml")}"

  vars {
    ecs_cluster_name = "${aws_ecs_cluster.container_instance.name}"
  }
}
```

`container-instance.yml` looks something like:

```yaml
#cloud-config

write_files:
  - path: /etc/ecs/ecs.config
    permissions: 0644
    owner: root:root
    content: |
      ECS_CLUSTER=${ecs_cluster_name}
```

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+}
+
+resource "aws_autoscaling_group" "ecs" {
+  lifecycle {
```
This lifecycle attribute should be removed.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+
+  # Explicitly linking ASG and launch configuration by name
+  # to force replacement on launch configuration changes.
+  name = "${aws_launch_configuration.ecs.name}"
```
RE: Naming, this should be something like `"asg${var.environment}ContainerInstance"`.

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  min_size                  = "${var.desired_instance_count}"
+  max_size                  = "${var.desired_instance_count}"
+  vpc_zone_identifier       = ["${var.subnet_id}"]
+
```
We should include an `enabled_metrics` attribute here that looks like:

```hcl
  enabled_metrics = [
    "GroupMinSize",
    "GroupMaxSize",
    "GroupDesiredCapacity",
    "GroupInServiceInstances",
    "GroupPendingInstances",
    "GroupStandbyInstances",
    "GroupTerminatingInstances",
    "GroupTotalInstances",
  ]
```

hectcastro commented on terraform/gt-site.tf:
```diff
@@ -0,0 +1,91 @@
+  max_size                  = "${var.desired_instance_count}"
+  vpc_zone_identifier       = ["${var.subnet_id}"]
+
+  tag {
```
We should try to include at least the follow three tags:

```hcl

  tag {
    key                 = "Name"
    value               = "ContainerInstance"
    propagate_at_launch = true
  }

  tag {
    key                 = "Project"
    value               = "${var.project}"
    propagate_at_launch = true
  }

  tag {
    key                 = "Environment"
    value               = "${var.environment}"
    propagate_at_launch = true
  }
```

hectcastro commented on terraform/state-persistence.tf:
```diff
@@ -0,0 +1,9 @@
+data "terraform_remote_state" "gt-site-cluster" {
```
This resource is setup to interact with Terraform state pushed and pulled primarily by another project. We should omit this resource and manage the Terraform state for this project via `terraform remote config` and `terraform remote push`:

```bash
            terraform remote config \
                      -backend="s3" \
                      -backend-config="region=us-east-1" \
                      -backend-config="bucket=${RF_SETTINGS_BUCKET}" \
                      -backend-config="key=terraform/state" \
                      -backend-config="encrypt=true"

...

                    terraform get -update
                    terraform plan \
                              -var-file="${RF_SETTINGS_BUCKET}.tfvars" \
                              -out="${RF_SETTINGS_BUCKET}.tfplan"

...

                    terraform apply "${RF_SETTINGS_BUCKET}.tfplan"
                    terraform remote push
```

fosskers replied:
> This resource is setup to interact with Terraform state pushed and pulled primarily by another project.

What do you mean, exactly? I changed the critical names to avoid clashes with the existing site state.

hectcastro replied:
Remote state for a given infrastructure is supposed to be pushed and pulled during each deployment via the [`remote`](https://www.terraform.io/docs/commands/remote.html) subcommand of `terraform`.

The `terraform_remote_state` data source is setup to be a read-only data source for scenarios where you're trying to extract references to resources created by someone else's Terraform driven infrastructure:

An example from the Terraform docs:

> For example, a core infrastructure team can handle building the core machines, networking, etc. and can expose some information to other teams to run their own infrastructure. As a more specific example with AWS: you can expose things such as VPC IDs, subnets, NAT instance IDs, etc. through remote state and have other Terraform states consume that.

fosskers replied:
Odd, I've noticed it pushing and pulling remote state via normal `terraform apply` and `terraform destroy`.

hectcastro replied:
Trying to read through some more docs and issues, it is still not entirely clear what this resource provides.

It does appear that remote state management has evolved a little. Apparently, after a `terraform remote config` invocation, `terraform` automatically pushes and pulls remote state.

I'm cool with leaving this is until we get to a point where I can take this PR for a spin, and then we'll be able to see how this remote state setup works on a machine that has no trace of existing Terraform state.

hectcastro commented on terraform/variables.tf:
```diff
@@ -0,0 +1,40 @@
+variable "profile" {
```
We should be able to remove this variable.

hectcastro commented on terraform/variables.tf:
```diff
@@ -0,0 +1,40 @@
+}
+
+variable "ecs_instance_profile" {
+  default = "arn:aws:iam::896538046175:instance-profile/terraform-wzxkyowirnachcosiqxrriheki"
```
This should be created via Terraform.

hectcastro commented on terraform/variables.tf:
```diff
@@ -0,0 +1,40 @@
+  default = "arn:aws:iam::896538046175:instance-profile/terraform-wzxkyowirnachcosiqxrriheki"
+}
+
+variable "ecs_service_role" {
```
Similar as above. This resource should be created via Terraform.

hectcastro commented on terraform/variables.tf:
```diff
@@ -0,0 +1,40 @@
+  default = "arn:aws:iam::896538046175:role/ecs_service_role"
+}
+
+variable "subnet_id" {
```
If we aren't planning to create a new VPC for this deployment, then we should be passing in multiple public and private subnet IDs. The public subnet IDs should be associated with `aws_alb` via `subnets` and private subnet IDs should be associated with `aws_autoscaling_group` via `vpc_zone_identifier`.

fosskers replied:
This is a bit beyond my understanding.

hectcastro replied:
The software defined network that is used to house the compute resources within an AWS account is referred to as a Virtual Private Cloud (VPC). VPCs have a private IP address space that is carved up into subnets, route tables, and gateways.

![image](https://cloud.githubusercontent.com/assets/43639/22959473/a8a30b28-f304-11e6-8cc4-cc1ded0434db.png)

This diagram from the VPC docs shows a single VPC that is carved up into two subnets (one public and one private). The public subnet contains EC2 instances with public IPs and a NAT device. The private subnet also contains EC2 instances, but those only have private IPs, and when they need to hit the internet, they need to be routed through the NAT in the public subnet.

In our case, we want:

- EC2 instances that are members of the ECS cluster to live within a private subnet
- ALB instances (transparent to us) in the public subnet

The `aws_alb` attribute `subnets` is how we associate an ALB with specific subnets. `vpc_zone_identifier` of `aws_autoscaling_group` is how we associate the same for ASGs.

I took a look at the default VPC in the GeoTrellis AWS account and it has a bunch of subnets, but the route tables are not setup to handle routing for private subnets that need to interact with the internet (no route table entries, no NAT device).

We usually lay down a new VPC for each deployment:

- https://github.com/azavea/raster-foundry-deployment/blob/develop/terraform/network.tf#L4-L19
- https://github.com/azavea/terraform-aws-vpc

In this case, we can opt to do that, or manually tweak the default VPC for the GeoTrellis AWS account so that it can support both public and private subnets.

fosskers replied:
Considering the site and demos are the only things GT deploys long-term, I think it makes more sense to tweak the GT AWS account as needed.

hectcastro replied:
Public subnets:

- `subnet-0b1b374d`
- `subnet-2ec98506`

Private subnets:

- `subnet-7d2ba926`
- `subnet-f2f078df`

hectcastro commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,26 @@
+# -*- mode: ruby -*-
```
I think that it would be good to keep this file at the root of the project. That is generally where it is kept across all other projects.

fosskers replied:
Sure, it used to be there but I foresaw you saying it should be relegated to its own directory for cleanliness. I can move it back.

hectcastro commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,26 @@
+Vagrant.configure("2") do |config|
+
+    # Ubuntu 16.04 LTS
+    config.vm.box = "fosskers/docker-1.12"
```
I think that we should be using a vanilla Vagrant box here vs. one that we have to maintain.

fosskers replied:
This was done to properly provision an image with docker without involving Ansible.

hectcastro replied:
I think that this approach can easily turn into a slippery slope. Right now, `fosskers/docker-1.12` is a bit of a black box.

- What process was used to create it?
- How can someone else update it? Report issues?
- What happens when the Docker Engine needs to be updated? What about Compose?

I'm not suggesting that the approach with Ansible is the ideal, but with that approach, all of these maintenance related questions have proven answers.

fosskers replied:
I should have mentioned this earlier: https://github.com/fosskers/vagrant-docker

Which can of course be moved to the Azavea org if needed. I should add a README explaining the box building + pushing process. At the moment it pushes to https://atlas.hashicorp.com/fosskers/boxes/docker-1.12.

fosskers replied:
This won't be needed if the Ansible provisioning problems (re: docker-compose) pan out.

hectcastro commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,26 @@
+
+    # VM resource settings
+    config.vm.provider :virtualbox do |vb|
+        vb.memory = 12288
```
This seems much higher than necessary. Maybe ~8GB?

fosskers replied:
8gb wasn't enough when I was doing ingests directly in Vagrant, but since we won't actually be doing that (any we're using an 8gb EC2 instance) this should be reduced from 12gb.

hectcastro commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    # Provisioning
+    config.vm.provision :shell, path: "bootstrap.sh"
```
I think that we should continue to use Ansible to provision Vagrant virtual machines vs. Bash scripts. We should also try to use Ansible to install Docker and its associated tooling.

fosskers replied:
Having never used Ansible, I wanted to avoid that learning overhead. The provisioning via bash is also fairly lean. 

We had our RRP this morning where I mentioned that Vagrant didn't feel entirely useful for our case here. Between a local `docker-compose` and spinning up a real cluster (quite quickly) with Terraform, Vagrant didn't feel like it had as strong a place in "proof of architecture correctness" as perhaps it does in other projects.

hectcastro replied:
I think the primary reason the Bash provisioning is lean is because a significant portion of it was done during the Vagrant box creation process. It is also not clear why all of the other dependencies are containerized, but the JDK still remains a dependency on the host.

I also disagree with the idea that Vagrant lacks value when interacting with the Docker Engine and Docker Compose. Sure, using Docker Compose from a Linux host is faster, but that tradeoffs comes with a bunch of cons as well.

- What version are you running? Is that version the same as everyone else on the team?
- Docker for Mac actually runs within a VM; what differences crop up when a developer workstation is Linux based vs. macOS based?
- Will it be the same the next time we touch this project in 6 months?
- How to we synchronize version bumps of the engine with ECS?

Again, all of these questions have proven answers when Vagrant is used to encapsulate development dependencies. Also, we got to this solution by enduring all of the cons associated with running Docker directly on people's workstations.

fosskers replied:
Using the Ansible provisioner would absolve us of having to upload things to the Hashicorp Atlas, so that's a plus. As for the overall value of Vagrant in this context, I've asked the team to weigh-in on this thread.

My general aversion to adding Ansible (or Vagrant at all) here comes from a desire to minimize moving parts with algorithms/systems.

lossyrob replied:
My thoughts are, we should be consistent with approach. The approach with other projects was to use ansible and vagrant, and I see there being a lot of worth in not breaking that pattern without a very clear benefit (which I don't see here).

I want to standardize in line with how our general deployment processes happen, and I'm working on creating some documentation around that. Until then I think it's wise to fall in line with what Hector suggests.

fosskers replied:
It turns out there was a built-in solution that didn't involves reams of extra configuration, as I had been originally (mistakenly) imagining.

hectcastro commented:
I'd also like to mention that since this repository is outside of the Azavea organization, email notifications get routed to my personal email account. That may lead to longer delays in response time because I don't check that as much during the day as my work email.

fosskers commented:
I fully intend to move it over to either `azavea` or `geotrellis` once this PR is in.

fosskers committed:
Move Vagrant config back out to project root

fosskers committed:
Provision Vagrant box with Ansible

- `docker-compose` doesn't work yet, this is being
investigated as I move to other subtasks.

fosskers committed:
(terraform) Make service healthy values parameters

- They have been given new default values as well, to avoid downtime.

fosskers committed:
(terraform) Use ALB over ELB

fosskers committed:
(terraform) More tags and `enabled_metrics`

fosskers committed:
(vagrant) Use `docker-compose 1.9.0`

- This avoids the strange pip/DC behaviour that appears with higher
versions.

fosskers committed:
(terraform) Remove `profile` variable

fosskers committed:
(terraform) Create IAM roles automatically

- VPC settings still unresolved.

fosskers committed:
(terraform) Add VPC id + security rule

fosskers committed:
(terraform) Make remote state file name dynamic

- Based on the `var.environment`.

fosskers committed:
(terraform) Remove old hard-coded ARN

fosskers committed:
(terraform) Use mix of public/private subnets

Commit:
Solidify Terraform support; serve website through CloudFront

Using STRTA `infra`, wrap the Terraform CLI so that remote state in S3 can be
pushed and pulled correctly via the updated Terraform 0.9 remote state support.

From there, add support for assembling all of the necessary AWS resources to run
the project within AWS via Amazon ECS. This consists of:

- An ECS cluster (managed via the `terraform-aws-ecs-cluster` module)
- An ECS service tied to an ALB for serving all traffic over HTTPS
- CloudWatch Log groups for service and container instance log aggregation
- A CloudFront distribution for redirecting HTTP traffic to HTTPS
- ALB and CloudFront log aggregation into an S3 bucket

hectcastro committed:
Merge pull request #2 from hectcastro/hmc/beta

Solidify Terraform support; serve website through CloudFront

tnation14 committed:
Update Vagrant project dependencies and folder structure

- Moves Terraform and Ansible resources into their own subfolders. The
  Vagrantfile and scripts/infra have been updated to reflect this change.

- Bump terraform version to 0.9.3, install awscli 11.82.* through pip.

tnation14 committed:
Remove all non-deployment resources.

Restructures repository to focus exclusively on deployment . Any
additional files will be removed and potentially added to other repos.

This commit:
- removes all git submodules
- removes Docker-compose setup, including docker-compose installation
- removes All STRTA besides `setup` and `infra`
- removes Website Terraform resources
- move all `deployment` subfolders to the root of the repo
- Updates Vagrantfile & `scripts/infra` to reflect directory structure changes

tnation14 committed:
Add travis config to run deployments

tnation14 committed:
Add VPC module

The existing GT Site infrastructure utilizes a VPC that is not managed by terraform, and uses VPC information in the form of variables. This PR adds a VPC using the `terraform-aws-vpc` module and threads through the VPC id to the container service.

tnation14 committed:
Add security group rules for the Bastion and Container instances

tnation14 committed:
Add DNS resources and record for Bastion instance

tnation14 committed:
Update path to remote state in scripts/infra

tnation14 committed:
Output remote state

Export information relevant to ECS tasks for use with other projects.

tnation14 committed:
Bump terraform version to 0.9.6

tnation14 committed:
Install AWS cli as a dependency.

tnation14 committed:
Fix after_deploy script syntax
